clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
Thanks for your patience with this issue, but I think it's really important that we think it all the way through. It actually occurs to me that one of the reasons this is turning out to be a little tricky is that the current code might not be handling all the cases around fetches correctly. I think the fact that `fetchedRecords` is not checked before blocking in poll() might already cause unnecessary blocking. The trouble is that the code seems to assume that fetches will only return in KafkaConsumer.poll(), but they could return in any blocking call (e.g. in a synchronous commit) since everything just delegates to NetworkClient.poll() in the end. That means that KafkaConsumer.poll() will block even when data is available. My proposal would probably be to leave Fetcher as it is in the current patch and add the check for non-empty records in KafkaConsumer.pollOnce() as I suggested before. Does that make sense?